### PR TITLE
http: implementing upstream filters

### PR DESCRIPTION
**Pull request by alyssawilk**
Major refactors to the upstream request to allow for upstream filters. This is runtime guarded but the refactor itself is high risk either way, due to the number of code points for data flow.

There's still a bunch of refactoring and clean up to do after this.  Upstream filters can do anything (except refresh route and do internal redirects) that downstream filters can, but have no ability to affect or access the upstream connection.  That will come as the upstream interfaces are added but I figured this was a large enough PR without that extra work.

Risk Level: high
Testing: LOADS of integration testing: all the filter integration tests now run for upstream/downstream.
Docs Changes: n/a (will once it's unhidden)
Release Notes: inline
Runtime guard: envoy_reloadable_features_allow_upstream_filters
Most of https://github.com/envoyproxy/envoy/issues/10455

### CODE REVIEW

**Comment by alyssawilk** [comment]
@RyanTheOptimist feel free to ignore this one for tonight - you already did a large PR for me today and I suspect it'll still fail coverage.  I'll work on unit tests in the morning but I'm still super stoked to have this "done"

File: changelogs/current.yaml [hunk 1]
```diff
@@ -59,6 +59,8 @@ new_features:
 - area: http
   change: |
     added default-false ``envoy.reloadable_features.http1_use_balsa_parser`` for experimental BalsaParser.
+  change: |
+    added ``envoy_reloadable_features_allow_upstream_filters`` for experimental upstream filters.
```
Comment (RyanTheOptimist) [inline changelogs/current.yaml:63]: Should this be `envoy.reloadable_features.allow_upstream_filters` instead of `envoy_reloadable_features_allow_upstream_filters`

File: source/common/http/conn_manager_impl.h [hunk 2]
```diff
@@ -279,7 +279,7 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     Router::RouteConstSharedPtr route(const Router::RouteCallback& cb) override;
     void setRoute(Router::RouteConstSharedPtr route) override;
     void clearRouteCache() override;
-    absl::optional<Router::ConfigConstSharedPtr> routeConfig() override;
+    absl::optional<Router::ConfigConstSharedPtr> routeConfig();
```
Comment (RyanTheOptimist) [inline source/common/http/conn_manager_impl.h:282]: nit: If this is no longer an override, we should probably move it out of the section of methods which all come from a subclass?

File: source/common/http/filter_manager.cc [hunk 3]
```diff
@@ -1335,7 +1335,9 @@ void FilterManager::setBufferLimit(uint32_t new_limit) {
 }
 
 void FilterManager::contextOnContinue(ScopeTrackedObjectStack& tracked_object_stack) {
-  tracked_object_stack.add(connection_);
+  if (connection_) {
```
Comment (RyanTheOptimist) [inline source/common/http/filter_manager.cc:1338]: OOC, is this a bug fix, or an artifact of adding upstream filters?

Comment (RyanTheOptimist) [reply source/common/http/filter_manager.cc:1338]: Oh, I guess it's because connection_ is now a pointer instead of a ref and it's a pointer that's allowed to be null? I think the Envoy idiom would be OptRef<Network::Connection> instead of a nullable pointer. Would that work?

All of THAT having been said, why is connection_ allowed to be null now? (This is probably obvious)

File: source/common/http/filter_manager.h [hunk 4]
```diff
@@ -1016,6 +1013,7 @@ class FilterManager : public ScopeTrackedObject,
     ActiveStreamDecoderFilter* latest_data_decoding_filter_{};
   };
 
+protected:
```
Comment (RyanTheOptimist) [inline source/common/http/filter_manager.h:1016]: Does this make state_ protected instead of private? If so, can we avoid that in some way? Protected members are not really style guide kosher.

File: source/common/router/router.cc [hunk 5]
```diff
@@ -805,8 +804,8 @@ Http::FilterTrailersStatus Filter::decodeTrailers(Http::RequestTrailerMap& trail
   // a backoff timer.
   ASSERT(upstream_requests_.size() <= 1);
   downstream_trailers_ = &trailers;
-  for (auto& upstream_request : upstream_requests_) {
-    upstream_request->acceptTrailersFromRouter(trailers);
+  if (!upstream_requests_.empty()) {
```
Comment (RyanTheOptimist) [inline source/common/router/router.cc:807]: OOC, why this change?

File: source/common/router/upstream_request.cc [hunk 6]
```diff
@@ -41,18 +41,190 @@
 namespace Envoy {
 namespace Router {
 
+// This is the last stop in the filter chain: take the headers and ship them to the codec.
```
Comment (RyanTheOptimist) [inline source/common/router/upstream_request.cc:44]: This is the start of a big block of newly added code. Is it copied from somewhere else, or is this all new?

File: source/common/router/upstream_request.h [hunk 7]
```diff
@@ -30,8 +31,52 @@ class GenericUpstream;
 class GenericConnectionPoolCallbacks;
 class RouterFilterInterface;
 class UpstreamRequest;
+class UpstreamRequestFilterManagerCallbacks;
+class UpstreamFilterManager;
+class CodecFilter;
 
-// The base request for Upstream.
+/* The Upstream request is the base class for forwarding HTTP upstream.
```
Comment (RyanTheOptimist) [inline source/common/router/upstream_request.h:38]: What an awesome comment! Thanks for writing it.

**Review by RyanTheOptimist — COMMENTED** [review]
There's a lot going on here, so only a few small comments.

(While I definitely feel encouraged to ignore this one for tonight... this is too exciting to ignore! :>)

File: source/common/http/conn_manager_impl.h (hunk 2 above)
Comment (alyssawilk) [reply source/common/http/conn_manager_impl.h:282]: good call

File: source/common/http/filter_manager.cc (hunk 3 above)
Comment (alyssawilk) [reply source/common/http/filter_manager.cc:1338]: so we can have filters on async client requests (no downstream)

File: source/common/http/filter_manager.h (hunk 4 above)
Comment (alyssawilk) [reply source/common/http/filter_manager.h:1016]: fixed

File: source/common/router/upstream_request.h (hunk 7 above)
Comment (alyssawilk) [reply source/common/router/upstream_request.h:38]: it'd be kinda cruel to make you figure it out on your own :-)

File: changelogs/current.yaml (hunk 1 above)
Comment (alyssawilk) [reply changelogs/current.yaml:63]: yep.  though this reminds me: @phlax in this instantiation of the PR the guard is true but even when it's false the scripts require comments.  I don't think we should require docs for guards which are false.

File: source/common/router/upstream_request.cc (hunk 6 above)
Comment (alyssawilk) [reply source/common/router/upstream_request.cc:44]: this is all new

**Comment by alyssawilk** [comment]
cc @paul-r-gall

**Comment by alyssawilk** [comment]
looks like an unrelated fuzz flake
/retest

File: source/common/router/upstream_request.cc [hunk 8]
```diff
@@ -72,6 +244,28 @@ UpstreamRequest::UpstreamRequest(RouterFilterInterface& parent,
   if (cluster_info.has_value()) {
     stream_info_.setUpstreamClusterInfo(*cluster_info);
   }
+  if (!allow_upstream_filters_) {
+    return;
+  }
+
+  // Set up the upstream filter manager.
+  fm_callbacks_ = std::make_unique<UpstreamRequestFilterManagerCallbacks>(*this);
```
Comment (RyanTheOptimist) [inline source/common/router/upstream_request.cc:252]: nit: I would call this filter_manager_callbacks_ to avoid abbreviations, but your call as that's pretty verbose.

File: source/common/router/upstream_request.cc [hunk 9]
```diff
@@ -588,6 +881,14 @@ void UpstreamRequest::encodeBodyAndTrailers() {
   }
 }
 
+UpstreamToDownstream& UpstreamRequest::upstreamToDownstream() {
+  if (allow_upstream_filters_) {
+    return codec_filter_->bridge_;
+  } else {
```
Comment (RyanTheOptimist) [inline source/common/router/upstream_request.cc:887]: nit: early return to avoid the else?

**Review by RyanTheOptimist — APPROVED** [review]
A couple of nits, but LGTM otherwise. This is awesome!

**Review by RyanTheOptimist — DISMISSED** [review]

**Review by RyanTheOptimist — DISMISSED** [review]

File: source/common/router/upstream_request.cc [hunk 10]
```diff
@@ -248,12 +465,40 @@ void UpstreamRequest::acceptHeadersFromRouter(bool end_stream) {
   router_sent_end_stream_ = end_stream;
 
   conn_pool_->newStream(this);
```
Comment (paul-r-gall) [inline source/common/router/upstream_request.cc:467]: Why don't we delay constructing a new stream from the connection pool until the upstream filter chain has been traversed? i.e. move this to the end of `acceptHeadersFromRouter`?

Comment (alyssawilk) [reply source/common/router/upstream_request.cc:467]: generally many streams are blocking, so kicking off the stream early would be important to many users.  I think we can make it configurable in the long run though.

**Comment by alyssawilk** [comment]
@mattklein123 ping?

**Comment by alyssawilk** [comment]
(working one morning this week from Poland, would love your comments by then)

File: source/common/router/upstream_request.h [hunk 11]
```diff
@@ -184,15 +236,191 @@ class UpstreamRequest : public Logger::Loggable<Logger::Id::router>,
...
+
+// This is the last filter in the upstream filter chain.
+// It takes request headers/body/data from the filter manager and encodes them to the upstream
+// codec. It also registers the CodecBridge with the upstream stream, and takes response
+// headers/body/data from the upstream stream and sends them to the filter manager.
+//
+// Right now, the CodecFilter is implicitly configured as the last filter in the
+// chain and manually created. TODO(alyssawilk) fix this before removing alpha
+// on upstream filters.
+class CodecFilter : public Http::StreamDecoderFilter, public Logger::Loggable<Logger::Id::router> {
```
Comment (soulxu) [inline source/common/router/upstream_request.h:371]: Just as the above comment said, this filter encodes headers/data/data to the upstream.
` It takes request headers/body/data from the filter manager and encodes them to the upstream codec.`

But why this filter is the subclass of `Http::StreamDecoderFilter`, not the `Http::StreamEncoderFilter`?

We take headers/body/data from the filter manager, then we should call `encodeHeader`, `encodeData`, etc...

File: source/common/router/upstream_request.cc [hunk 12]
```diff
@@ -72,6 +244,28 @@ UpstreamRequest::UpstreamRequest(RouterFilterInterface& parent,
...
+      parent_.callbacks()->connection(), parent_.callbacks()->streamId(),
+      parent_.callbacks()->account(), true, parent_.callbacks()->decoderBufferLimit(),
+      *parent_.cluster(), *this);
+  parent_.cluster()->createFilterChain(*filter_manager_);
+
+  // Similar to how the downstream filter chain ends with the router, the
+  // upstream filter chain ends with the encoder filter, which sends data to the
+  // codec. Create that manually here.
+  Http::FilterFactoryCb cb = [this](Http::FilterChainFactoryCallbacks& callbacks) -> void {
+    codec_filter_ = std::make_shared<CodecFilter>(*this);
```
Comment (soulxu) [inline source/common/router/upstream_request.cc:264]: a question on why we didn't pass the `Http::FilterChainFactoryCallbacks` to the filter instead of UpstreamRequest object, I thought `UpstreamRequest` is Envoy internal implement, shouldn't expose to the filter.

File: source/common/router/upstream_request.h [hunk 13]
```diff
@@ -130,6 +180,9 @@ class UpstreamRequest : public Logger::Loggable<Logger::Id::router>,
   bool hadUpstream() const { return had_upstream_; }
 
 private:
+  friend class UpstreamFilterManager;
+  friend class CodecFilter;
```
Comment (soulxu) [inline source/common/router/upstream_request.h:184]: Emm...this is related to my question at https://github.com/envoyproxy/envoy/pull/22553/files/e755fee044c1ca990736a3ed605dfb729d6052d4#r945441918

Seems the filter shouldn't reference to the UpstreamRequest, it should use the filter manager callback.

File: source/common/router/upstream_request.h (hunk 11 above)
Comment (soulxu) [reply source/common/router/upstream_request.h:371]: I guess the reason is we defined `Http::StreamDecoderFilter` for the requesting process, `Http::StreamEncoderFIlter` for the response process. 

But It just feels strange put the word decoder/encoder in the upstream part.

File: source/common/router/upstream_request.cc (hunk 12 above)
Comment (soulxu) [reply source/common/router/upstream_request.cc:264]: I think I get answer by my self https://github.com/envoyproxy/envoy/pull/22553/files/e755fee044c1ca990736a3ed605dfb729d6052d4#diff-889698b89a1023b7d44c1c6a715e0b05b7d92c399e0104998290ba4a605fadfdR73-R77, you plan to clean up this later :)

**Comment by mattklein123** [comment]
> @mattklein123 ping?

Sorry I didn't realize this was waiting on me. I will review today.

File: source/common/router/upstream_request.cc [hunk 14]
```diff
@@ -248,12 +465,40 @@ void UpstreamRequest::acceptHeadersFromRouter(bool end_stream) {
...
+  }
+
+  // Make sure that when we are forwarding CONNECT payload we do not do so until
+  // the upstream has accepted the CONNECT request.
+  if (headers->getMethodValue() == Http::Headers::get().MethodValues.Connect) {
+    paused_for_connect_ = true;
+  }
 
+  filter_manager_->requestHeadersInitialized();
+  //  filter_manager_.setDownstreamRemoteAddress(mutate_result.final_remote_address);
```
Comment (mattklein123) [inline source/common/router/upstream_request.cc:496]: ?

File: source/common/router/upstream_request.h [hunk 15]
```diff
@@ -30,8 +31,52 @@ class GenericUpstream;
...
+ *
+ * This supports both classic forwarding, and upstream filter mode, dictated by
+ * allow_upstream_filters_ which is set via envoy.reloadable_features.allow_upstream_filters.
+ *
+ * In 'classic' mode, the UpstreamRequest requests a new stream, internally
+ * buffers data and trailers in |buffered_request_body_| and |downstream_metadata_map_vector_|
+ * and forwards the router's headers, buffered body, buffered metadata, and
+ * router's trailers to the |upstream_| once the stream is established.
+ *
+ * In 'classic; mode, when the stream is established, the UpstreamRequest
```
Comment (mattklein123) [inline source/common/router/upstream_request.h:48]:
```suggestion
 * In 'classic' mode, when the stream is established, the UpstreamRequest
```

File: source/common/router/upstream_request.cc (hunk 10 above)
Comment (mattklein123) [reply source/common/router/upstream_request.cc:467]: I think this is fine, but might be worth a comment here on the implication of this. I have a vague worry that this might cause issues for idle timeouts or something if there are filters that pause the request flow for a while. I guess in general all upstream timeouts would be impacted due to filters so probably not a big deal, but might be worth a comment/TODO here either way.

File: changelogs/current.yaml [hunk 16]
```diff
@@ -62,6 +62,8 @@ new_features:
 - area: http
   change: |
     added default-false ``envoy.reloadable_features.http1_use_balsa_parser`` for experimental BalsaParser.
+  change: |
+    added ``envoy.reloadable_features.allow_upstream_filters`` for experimental upstream filters.
```
Comment (mattklein123) [inline changelogs/current.yaml:66]: Can we get some stuff added to arch overview somewhere? Fine if you want to do this as a follow up if people are eager to land this. I'm thinking we should talk about this in the general HTTP section and maybe have a configuration overview sample for some upstream use case? With all the existing filters I'm sure we can come up with something?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this is super cool. Amazing work on such a long requested feature! :tada: 

TBH I mostly skimmed the logic at a high level. As you say, this is super high risk, and I think the chance of catching all logic issues during code review is zero so we might as well ship and get some test coverage and :crossed_fingers:. For this change it might be nice to get someone to canary it first but up to you if you want to go to that effort. 

Thank you!

File: source/common/router/router.cc [hunk 17]
```diff
@@ -805,8 +804,8 @@ Http::FilterTrailersStatus Filter::decodeTrailers(Http::RequestTrailerMap& trail
...
+    upstream_requests_.front()->acceptTrailersFromRouter(trailers);
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/common/router/router.cc:808]: yep - we switched all the other loops but this one and if the connection goes down during handling one stream it's problematic.

File: source/common/router/upstream_request.cc (hunk 6 above)
Comment (alyssawilk) [reply source/common/router/upstream_request.cc:44]: I don't think there's much more code sharing that can be done.
Last filter in the downstream chain is the router, and that's pretty different from the codec.

File: source/common/router/upstream_request.cc (hunk 12 above)
Comment (alyssawilk) [reply source/common/router/upstream_request.cc:264]: so here I would like to move over to the HCM creating the filter chain, which will be shared code.  I've got a TODO for that already insofar as it can't be done until all the direct links between codec filter and upstream request go away.  Got 90% of that refactor done in the next follow-up :-)

File: source/common/router/upstream_request.h (hunk 13 above)
Comment (alyssawilk) [reply source/common/router/upstream_request.h:184]: Agreed!  the filter ideally wouldn't reference the upstream request, but fixing that requires giving upstream filters access to an interface which allows them upstream-specific information.  I've got a TODO to clean up the codec filter so it can be created by the factory, and to accomplish that I have to remove all direct access, and can remove this friend class.    It's a lot more complicated code so I planned on adding it as a follow up.

File: source/common/router/upstream_request.h (hunk 11 above)
Comment (alyssawilk) [reply source/common/router/upstream_request.h:371]: because decode filters handle request path.  encode filters _alter_ the response path. This is essentially the same  as the router filter which is a decide only filter.  but yeah agree the encode/decode codec vs encode/decode filter is confusing.

**Review by mattklein123 — DISMISSED** [review]
LGTM! Sweet!

**Comment by alyssawilk** [comment]
@mattklein123 can I get a stamp on the merge resolution?

**Review by mattklein123 — APPROVED** [review]
